demos/fleet_provisioning/fleet_provisioning_with_csr/fleet_provisioning_payload_operations.c

```diff
@@ -0,0 +1,372 @@
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+/* TinyCBOR library for CBOR encoding and decoding operations. */
+#include "cbor.h"
+
+/* Demo config. */
+#include "demo_config.h"
+
+/* AWS IoT Fleet Provisioning Library. */
+#include "fleet_provisioning.h"
+
+/* Header include. */
+#include "fleet_provisioning_payload_operations.h"
+/*-----------------------------------------------------------*/
+
+bool generateCsrRequest( char * pBuffer,
+                         size_t bufferLength,
+                         const char * pCsr,
+                         size_t csrLength,
+                         size_t * pOutLengthWritten )
+{
+    bool status = false;
+    CborEncoder encoder, mapEncoder;
+    CborError cborRet;
+
+
+    /* For details on the CreateCertificatefromCsr request payload format, see:
+     * https://docs.aws.amazon.com/iot/latest/developerguide/fleet-provision-api.html#create-cert-csr-request-payload
+     */
+    if( status == true )
+    {
+        cbor_encoder_init( &encoder, ( uint8_t * ) pBuffer, bufferLength, 0 );
+
+        /* The request document is a map with 1 key value pair. */
+        cborRet = cbor_encoder_create_map( &encoder, &mapEncoder, 1 );
+
+        if( cborRet == CborNoError )
+        {
+            cborRet = cbor_encode_text_stringz( &mapEncoder, "certificateSigningRequest" );
+        }
+
+        if( cborRet == CborNoError )
+        {
+            cborRet = cbor_encode_text_string( &mapEncoder, pCsr, csrLength );
+        }
+
+        if( cborRet == CborNoError )
+        {
+            cborRet = cbor_encoder_close_container( &encoder, &mapEncoder );
+        }
+
+        if( cborRet == CborNoError )
+        {
+            *pOutLengthWritten = cbor_encoder_get_buffer_size( &encoder, ( uint8_t * ) pBuffer );
+        }
+        else
+        {
+            status = false;
+            LogError( ( "Error during CBOR encoding: %s", cbor_error_string( cborRet ) ) );
+
+            if( ( cborRet & CborErrorOutOfMemory ) != 0 )
+            {
+                LogError( ( "Cannot fit CreateCertificateFromCsr request payload into buffer." ) );
+            }
+        }
+    }
+
+    return status;
```

Comment:
The use of `status` can be avoided to use a single status carrying variable as:
```suggestion
    return ( cborRet == CborNoError );
```